test/Test/Data/Mutable/Array.hs

```diff
@@ -0,0 +1,230 @@
+{-# LANGUAGE ApplicativeDo #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+
+-- |
+-- Tests for mutable arrays.
+--
+-- The API is basically 4 mutators or constructors: alloc, write, resize,
+-- resizeSeed and 2 accessors: length, read.
+--
+-- We test arrays behave properly with properties that axiomatically define
+-- arrays.  Basically, using these properties, the behavior of any use of
+-- arrays should be fully specified without having to look at the
+-- implementation.
+--
+-- An example of such a property would be the following:
+--
+-- for all i, j, arr, s.t. i /= j and 0 <= i,j < (length arr),
+--             read (write arr j x) i == read arr i
+--
+-- In general, most uses of arrays are of the form
+-- \[  accessor (constructor (...)) \].
+--
+-- Hence, we define properties that allow you to "simplyfy" or "rewrite"
+-- any such combination into a simpler form. (More formally,
+-- we're defining term re-write rules.) For now, constructor-constructor
+-- re-writes are skipped (e.g., write (write a i x) i y = write a i y).
+--
+-- TODO:
+--  * Test failures for out of bound access
+--  * Constructor - constructor rules, like
+--                write (write a i x) i y = write a i y
+module Test.Data.Mutable.Array
+  ( mutArrTests,
+  )
+where
+
+import qualified Data.Array.Mutable.Linear as Array
+import Data.Unrestricted.Linear
+import Hedgehog
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import qualified Prelude.Linear as Linear
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.Hedgehog (testProperty)
+
+-- # Exported Tests
+--------------------------------------------------------------------------------
+
+mutArrTests :: TestTree
+mutArrTests = testGroup "Mutable array tests" group
+
+group :: [TestTree]
+group =
+  -- All tests for exprs of the form (read (const ...) i)
+  [ testProperty "∀ s,i,x. read (alloc s x) i = x" readAlloc
+  , testProperty "∀ s,a,i. read (resize s a) i = read a 0" readResize
+  , testProperty "∀ a,i,x. read (write a i x) i = x " readWrite1
+  , testProperty "∀ a,i,j/=i,x. read (write a j x) i = read a i" readWrite2
+  , testProperty "∀ a,s,x,i. read (resizeSeed s x a) i = x" readResizeSeed
+  -- All tests for exprs of the form (length (const ...))
+  , testProperty "∀ s,x. len (alloc s x) = s" lenAlloc
+  , testProperty "∀ a,i,x. len (write a i x) = len a" lenWrite
+  , testProperty "∀ a,s. len (resize s a) = s" lenResize
+  , testProperty "∀ a,s,x. len (resizeSeed s x a) = s" lenResizeSeed
+  ]
+
+-- # Internal Library
+--------------------------------------------------------------------------------
+
+type ArrayTester = Array.Array Int #-> Unrestricted (TestT IO ())
+
+-- | Given a size, make a random list of that lenght
+randomList :: Int -> Gen [Int]
+randomList size = do
```

Comment:
Prefer

```suggestion
list :: Int -> Gen [Int]
list size = do
```

Why? because, then you will write

```haskell
x <- forAll list
```

The point is that the property is not about randomness, or random lists, it's about lists. All of them. It happens to be checked using random examples, but that's all.

Reply:
Oh -- yes, of course. That's brilliant and makes a lot of sense :)